assets/js/hooks/hoverune.js

```diff
@@ -70,16 +62,19 @@ export default HoveRune = {
           return;
         }
 
-        var getSelectRect = selection.getRangeAt(0).getBoundingClientRect();
+        var range = selection.getRangeAt(0);
+        var getSelectRect = range.getBoundingClientRect();
```

Comment:
```suggestion
        var getSelectRect = <something else??.getBoundingClientRect();
```

I don't think we should merge this version of the ui. 
Consider this after selection vs before selection case. 


after selection
![image](https://github.com/user-attachments/assets/2bf507aa-56a0-4b85-8538-5aec9ceeb741)

before selection: 
![image](https://github.com/user-attachments/assets/90868600-cefb-49bf-8348-4fbac5ef90fb)

#### my suggestion:
You should use the floater hook that I had defined earlier, it's for this usecase. 


```js
Floater = {
  mounted() {
    console.log("[floater] floater mounted");
    console.log("[floater] dataset check: ", this.el.dataset);
    const { floaterId, floaterReferenceSelector } = this.el.dataset;
  },
  beforeUpdate() {
    // gets called synchronously, prior to update
    console.log("[floater] Triggerred floater::beforeUpdate()");
    const { floater, reference, fallback } = this.getRelevantElements();

    // TODO: this is hardcoded to the media bridge, refactor when more sane.
    const offsetHeight = fallback.offsetHeight; // so pretend it's lower by this amount
    const isReferenceOutOfView = isElementOutOfViewport(reference, {
      top: 0,
      bottom: offsetHeight,
      left: 0,
      right: 0,
    });
    if (isReferenceOutOfView) {
      console.log(
        "[floater] Reference is out of viewport, should use fallback",
        {
          floater,
          reference,
          fallback,
        },
      );
    }
    const target =
      isMobileDevice() || isReferenceOutOfView ? fallback : reference;
    this.alignFloaterToRef(floater, target);
  },
  updated() {
    // gets called when the elem changes
    console.log("[floater] Triggerred floater::updated()");
    const { floater, reference, fallback } = this.getRelevantElements();
  },
``` 


---- 

ref perplexity outline on how to potentially do this

<details><summary>Ref perplexity outline on how to potentially do this</summary>


<details><summary>My Query</summary>
<p>


on an elixir liveview app, i'm trying to use the floater.js package 

i have a hook defined like so:


```js
/*
 * Ideally generic hook for floating logic.
 */
import { isMobileDevice } from "../utils/device.js";
import {
  computePosition,
  autoPlacement,
  shift,
  offset,
} from "floating-ui.dom.umd.min";

Floater = {
  mounted() {
    console.log("[floater] floater mounted");
    console.log("[floater] dataset check: ", this.el.dataset);
    const { floaterId, floaterReferenceSelector } = this.el.dataset;
  },
  beforeUpdate() {
    // gets called synchronously, prior to update
    console.log("[floater] Triggerred floater::beforeUpdate()");
    const { floater, reference, fallback } = this.getRelevantElements();

    // TODO: this is hardcoded to the media bridge, refactor when more sane.
    const offsetHeight = fallback.offsetHeight; // so pretend it's lower by this amount
    const isReferenceOutOfView = isElementOutOfViewport(reference, {
      top: 0,
      bottom: offsetHeight,
      left: 0,
      right: 0,
    });
    if (isReferenceOutOfView) {
      console.log(
        "[floater] Reference is out of viewport, should use fallback",
        {
          floater,
          reference,
          fallback,
        },
      );
    }
    const target =
      isMobileDevice() || isReferenceOutOfView ? fallback : reference;
    this.alignFloaterToRef(floater, target);
  },
  updated() {
    // gets called when the elem changes
    console.log("[floater] Triggerred floater::updated()");
    const { floater, reference, fallback } = this.getRelevantElements();
  },
  alignFloaterToRef(floater, reference) {
    const canBeAligned = floater && reference;
    if (!canBeAligned) {
      console.log("[floater] Can't be aligned");
      return;
    }

    computePosition(reference, floater, {
      placement: "right",
      // NOTE: order of middleware matters.
      middleware: [
        autoPlacement({
          allowedPlacements: ["right", "top-end"],
        }),
        shift({
          padding: 8,
          crossAxis: true,
        }),
        offset(6),
      ],
    }).then(({ x, y }) => {
      console.log("[floater] computed coordinates:", { x, y });
      Object.assign(floater.style, {
        left: `${x}px`,
        top: `${y}px`,
      });
    });
  },
  getRelevantElements() {
    const {
      floaterId,
      floaterReferenceSelector,
      floaterFallbackReferenceSelector,
    } = this.el.dataset;
    const floater = document.getElementById(floaterId);
    const reference = document.querySelector(floaterReferenceSelector);
    const fallback = document.querySelector(floaterFallbackReferenceSelector);

    console.log("[floater] getRelevantElements", {
      floater,
      reference,
      fallback,
      isMobileDevice: isMobileDevice(),
    });

    return {
      floater,
      reference,
      fallback,
    };
  },
};

// offset: more positive is more in that direction. so if left = 2 vs left = 3, then the second left is more left than the first left lol.
// offest is to be applied to the value of the rect so rect with offset top = 2 is as though the original left +2 in height
function isElementOutOfViewport(
  el,
  offsets = { top: 0, bottom: 0, left: 0, right: 0 },
) {
  if (!el) {
    console.log("[floater] el is null", el);
  } else {
    const rect = el.getBoundingClientRect();
    const { top, bottom, left, right } = offsets;

    return (
      rect.top + top < 0 ||
      rect.left + left < 0 ||
      rect.bottom + bottom >
        (window.innerHeight || document.documentElement.clientHeight) ||
      rect.right + right >
        (window.innerWidth || document.documentElement.clientWidth)
    );
  }
}

export default Floater;

``` 


here's an example of when it's used


  def video_player(assigns) do
    ~H"""
    <div>
      <div
        class={
          if @should_show_vid, do: "container-YouTubePlayer", else: "container-YouTubePlayerHidden"
        }
        id="container-YouTubePlayer"
        heartbeat={@heartbeat}
        phx-hook="Floater"
        data-floater-id="container-YouTubePlayer"
        data-floater-reference-selector=".emphasized-verse"
        data-floater-fallback-reference-selector="#media-player-container"
      >
        <.live_component
          module={VyasaWeb.YouTubePlayer}
          id="YouTubePlayer"
          video_id={@video.ext_uri}
          player_config={@player_config}
        />
      </div>
    </div>
    """
  end

what i want is to be able to spawn a function component called FOO and i want it to be based on my cursor selection of any text that I do on my screen via the default text selection capability of a browser


some questions:
1. can i target location of currently selected text and use that as the reference argument?
2. is there any other way i can insert hidden divs or hidden spans so that I can get this to work?
3. you may suggest improvements to the generic hook i've defined above and try to improve its interface
</p>
</details> 


<details><summary>Answer</summary>
<p>


To implement a feature in your Elixir LiveView application where you can use the `Floater.js` package to show a floating component based on the currently selected text, here are the answers to your questions along with some suggestions for improving your hook.

### 1. Targeting the Location of Currently Selected Text

You can indeed target the location of the currently selected text in the browser. The `window.getSelection()` method allows you to retrieve the current selection, and from there, you can get the bounding rectangle of the selection to position your floater. Here’s how you can do it:

```javascript
getSelectionBoundingRect() {
  const selection = window.getSelection();
  if (selection.rangeCount > 0) {
    const range = selection.getRangeAt(0);
    return range.getBoundingClientRect();
  }
  return null;
}
```

You can then use this bounding rectangle to align your floater:

```javascript
const rect = this.getSelectionBoundingRect();
if (rect) {
  // Use rect to position your floater
  this.alignFloaterToRef(floater, rect);
}
```

### 2. Inserting Hidden Divs or Spans

If you want to insert hidden elements dynamically, you can create a `<div>` or `<span>` and set its `display` style to `none`. You can then show it when needed. Here’s an example:

```javascript
const hiddenDiv = document.createElement('div');
hiddenDiv.style.display = 'none';
hiddenDiv.id = 'foo-component';
document.body.appendChild(hiddenDiv);

// Later, when you want to show it:
hiddenDiv.style.display = 'block';
```

### 3. Improvements to the Generic Hook

Here are some suggestions for improving your existing `Floater.js` hook:

- **Dynamic Positioning Based on Selection**: As mentioned, incorporate logic to position based on text selection.
  
- **Cleanup Logic**: Add a cleanup function that removes any event listeners or dynamically created elements when the floater is no longer needed.

- **Debouncing Updates**: If you're frequently updating the position of the floater, consider debouncing these updates to improve performance.

- **Accessibility Improvements**: Ensure that your floater is accessible. This could include adding ARIA roles and properties.

Here’s an improved version of your hook that includes dynamic positioning based on text selection:

```javascript
Floater = {
  mounted() {
    console.log("[floater] floater mounted");
    this.updateFloaterPosition(); // Initial positioning
    document.addEventListener('selectionchange', () => this.updateFloaterPosition());
  },
  
  updateFloaterPosition() {
    const { floater } = this.getRelevantElements();
    const rect = this.getSelectionBoundingRect();
    if (rect) {
      this.alignFloaterToRef(floater, rect);
    }
  },

  getSelectionBoundingRect() {
    const selection = window.getSelection();
    if (selection.rangeCount > 0) {
      const range = selection.getRangeAt(0);
      return range.getBoundingClientRect();
    }
    return null;
  },

  alignFloaterToRef(floater, reference) {
    const canBeAligned = floater && reference;
    if (!canBeAligned) {
      console.log("[floater] Can't be aligned");
      return;
    }

    computePosition(reference, floater, {
      placement: "right",
      middleware: [
        autoPlacement({ allowedPlacements: ["right", "top-end"] }),
        shift({ padding: 8, crossAxis: true }),
        offset(6),
      ],
    }).then(({ x, y }) => {
      Object.assign(floater.style, {
        left: `${x}px`,
        top: `${y}px`,
      });
    });
  },

  // ... rest of your existing methods remain unchanged
};
```

### Conclusion

With these adjustments, you should be able to effectively use the `Floater.js` package in your Elixir LiveView application to create a floating component based on user text selections. This approach enhances user interaction by making it contextually relevant and visually appealing. If you have further questions or need additional help implementing these features, feel free to ask!

Citations:
[1] https://alexcican.com/post/hide-and-show-div/
[2] https://www.geeksforgeeks.org/how-to-read-all-spans-of-a-div-dynamically/
[3] https://stackoverflow.com/questions/3181162/hide-span-or-div-until-javascript-completes/3181189
[4] https://forums.phpfreaks.com/topic/315528-simple-show-hide-html-span-solution/
[5] https://allyjs.io/tutorials/hiding-elements.html
[6] https://stackoverflow.com/questions/77996195/hide-and-show-span-element-by-clicking-div-element
[7] https://www.shecodes.io/athena/5942-how-to-use-text-overflow-hidden-with-a-span-tag
[8] https://gsap.com/community/forums/topic/35179-gsap-splittext-wraps-characters-in-divs-instead-of-span/
</p>
</details> 


</details>